numpy/lib/function_base.py

```diff
@@ -1574,17 +1609,38 @@ def gradient(f, *varargs, **kwargs):
 
     n = len(varargs)
     if n == 0:
-        dx = [1.0]*N
+        dx = [1.0] * N
     elif n == 1:
-        dx = [varargs[0]]*N
+        if np.isscalar(varargs[0]):
+            dx = [varargs[0]] * N
+        else:
+            if(len(varargs[0]) != f.shape[0]) or len(axes) != 1:
+                raise ValueError("distances must be either scalars or match "
+                                 "the length of the corresponding dimension")
+            diffx = np.diff(varargs[0])
+            # if distances are constant reduce to the scalar case
```

Comment:
uhm,  IMHO it doesn't add too much complexity and may help save a bit of memory, but ok I'll remove it